auto-instrumentation: propagate detected langs to injector

stanistan opened the pull request:
### What does this PR do?

https://datadoghq.atlassian.net/browse/APMON-1364

There is a possible mismatch between _injector_ language detection and _process agent_ language detection. This PR propagates the languages detected (while not changing container/injection behavior) through to the application container in environment variables.

- `DD_INSTRUMENTATION_LANGUAGES_DETECTED`
- `DD_INSTRUMENTATION_LANGUAGE_DETECTION_INJECTION_ENABLED`



<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Pod-startup time is something we are trying to control with the auto-instrumentation, and `language_detection` (which we have yet to enable for the controller), is one method of making this better. The risk of turning on language detection lies with us having having potentially inconsistent results. So we're passing the environment variables to the injector so it can deal with that on its own/directly.

### Additional Notes

1. We send back the detected languages _alongwith_ the ones we will use.
2. extract out `libInfoSource` which has `autodetected() bool` and `injectionType() string`. This also removes redundant logic for setting up `DD_INSTRUMENTATION_INSTALL_TYPE`, which was set twice (first one taking precedence) and added tests for this to be set.
3. `libInfoSource` is a `podMutator`
4. `detectedLanguages` is a mutator
5. `extractLibInfo` returns a struct instead of multiple returns so it's easeir to test, function and pass around

### Possible Drawbacks / Trade-offs

N/A

### Describe how to test/QA your changes

Change is covered with unit tests and testing using https://github.com/DataDog/k8s-ssi-v2-testing/

<img width="462" alt="Screenshot 2024-08-13 at 7 17 50 PM" src="https://github.com/user-attachments/assets/ba8a8ce8-eb00-4cb6-a20e-000f930fb65b">

stanistan committed:
auto-instrumentation: propagate detected langs

https://datadoghq.atlassian.net/browse/APMON-1364

1. We send back the detected langauges _with_ the ones we will use.
2. extract out `libInfoSource` which has `autodetected() bool` and
   `injectionType() string`. This also removes redundant logic for
   setting up `DD_INSTRUMENTATION_INSTALL_TYPE`, which was set twice
   (first one taking precedence) and added tests for this to be set.
3. `libInfoSource` is a `podMutator`
4. `detectedLanguages` is a mutator
5. `extractLibInfo` returns a struct instead of multiple returns
   so it's easeir to test, function and pass around

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -202,19 +202,7 @@ func (w *Webhook) inject(pod *corev1.Pod, ns string, _ dynamic.Interface) (bool,
 	}
 
-	// Inject env variables used for Onboarding KPIs propagation
-	var injectionType string
```
This was being done twice, once in `injectAutoInstruConfig` -- already covered by existing tests for `v1` and added more tests for this in `v2`.

stanistan committed:
fix source attribution for telemetry

stanistan committed:
always propagate detected langs if they are present

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_util.go:
```diff
@@ -50,6 +50,12 @@ func getLibListFromDeploymentAnnotations(store workloadmeta.Component, deploymen
 		for lang := range languages {
+			// There's a mismatch between language detection and auto-instrumentation.
+			// The Node language is a js lib.
+			if lang == "node" {
```
Found this in testing where if lang-detected "node" we would try to inject "node", but we don't have that in the webhook: 

https://github.com/DataDog/datadog-agent/blob/fe13deade9ef6cf2444c8c5e5284c7ef89f6794d/pkg/clusteragent/admission/mutate/autoinstrumentation/language_versions.go#L18-L25

jonbodner replied:
From taking a look at the auto-inject code, I think the name that telemetry expects is `nodejs`. Should we be consistent?

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -450,25 +530,35 @@ func (w *Webhook) newInjector(startTime time.Time, pod *corev1.Pod) *injector {
 	}
 
-	// inject the env for the pod.
-	_ = mutatecommon.InjectEnv(pod, localLibraryInstrumentationInstallTypeEnvVar)
```
replaced by `config.source.mutatePod(..)` below.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_util_test.go:
```diff
@@ -71,7 +71,6 @@ func TestGetOwnerNameAndKind(t *testing.T) {
 			require.Equal(t, found, tt.wantFound)
 			require.Equal(t, name, tt.expectedName)
 			require.Equal(t, kind, tt.expectedKind)
-
```
nit: unnecessary line removal

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +296,47 @@ func getPinnedLibraries(registry string) []libInfo {
+	})
+}
+
+// getLibrariesLanguageDetection runs process language auto-detection and returns languages
```
small nit:

This part doesn't run process language auto-detection.

Process language auto-detection runs in the process check/process collector. 

A more accurate description is:

`getLibrariesLanguageDetection returns the languages that were detected by process language detection`

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +296,47 @@ func getPinnedLibraries(registry string) []libInfo {
-	}
+type libInfoLanguageDetection struct {
+	libs    []libInfo
+	enabled bool
```
IMHO, I think `enabled` can be misleading because when we read `libInfoLanguageDetection.enabled` we might think that this indicates if language detection is enabled or not, which is not the case. 

I assume this property refers to whether library injection is based on language detection or not, so maybe `injectionEnabled` might be clearer about the intent.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +296,47 @@ func getPinnedLibraries(registry string) []libInfo {
+			Value: strings.Join(libs, ","),
+		})
+		_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+			Name:  "DD_INSTRUMENTATION_LANGUAGE_DETECTION_ENABLED",
```
There is a distinction between "enabling language detection" and "enabling injection based on language detection". 

enabling language detection means running the process check/collector to detect processes and languages. This is controlled by `datadog.apm.instrumentation.language_detection.enabled` helm config flag. 

enabling injection based on language detection means using the results of auto-detection to inject only the needed libraries in the admission controller. This is controlled by `admission_controller.auto_instrumentation.inject_auto_detected_libraries` 

Given this, it is important to be clear on what we are passing down to the injector. Do we need to indicate both? If yes, the code should be adapted to do so. If not, which one is important to be passed down? 

With the current implementation in this PR, I see that we are indicating if **injection based on language detection** is enabled or not by embedding a boolean flag as an env var `DD_INSTRUMENTATION_LANGUAGE_DETECTION_ENABLED`. But the env var might be misleading and fool us to think that it indicates whether or not language detection is enabled.

adel121 replied:
A small indication, to check if language detection is enabled or not, you can assert that `language_detection.enabled` and `language_detection.reporting.enabled` are both set to `true`

stanistan replied:
I'm changing this to `DD_INSTRUMENTATION_LANGUAGE_DETECTION_INJECTION_ENABLED` so it's clearer

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +349,94 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+
+const (
+	libInfoSourceNone libInfoSource = iota
+	libInfoSourceLibInjection
+	libInfoSourceSingleStepInstrumentation
+	libInfoSourceSingleStepLangaugeDetection
```
Can you please explain what each of these refer to? 

I might be able to guess, but want to make sure I understand correctly, and maybe we can improve the naming so that the names convey the intent better.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +349,94 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+		}
+	)
+
+	if ssiEnabled {
+		extracted.source = libInfoSourceSingleStepInstrumentation
+	}
+
```
if my understanding is correct, `ssiEnabled` will be `true` whener the namespace is enabled for APM SSI. 

It means that always in this case we will have `extracted.source = libInfoSourceSingleStepInstrumentation` no matter if the results were coming from pinned libraries or from process language detection.

stanistan replied:
Yes, this is consistent with the old behavior (and there are tests for it). I can fix it here (I think this might be a bug) where if we're using the annotations we should be doing `libInjection`

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +349,94 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+// injecting into the pod and where it came from.
+func (w *Webhook) extractLibInfo(pod *corev1.Pod) extractedPodLibInfo {
+	var (
+		ssiEnabled        = w.isEnabledInNamespace(pod.Namespace)
```
I think `ssiEnabled` as a name doesn't really reflect the intent.

If single step instrumentation was disabled, this webhook wouldn't be registered at the first place, and we won't reach this point. 

Maybe `namespaceEnabled` is a better name?

adel121 replied:
I also don't think we will reach this function if the namespace was not enabled, which means that here `ssiEnabled` will never be false.

Or am I missing something?

stanistan replied:
We do, `isPodEligible` is a superset of `isNamespaceEnabled`, you can get here if you've opted in or out of injection injection via the annotation `admission.datadoghq.com/enabled`, the ssi flag, or `mutate_unlabeled` configs. 

So `ssi` specifically is only _one_ of the factors where we disambiguate between "lib injection" (where someone specified annotations to get this data) and "single step" where `apm.instrumentation.enabled == true`

stanistan replied:
>If single step instrumentation was disabled, this webhook wouldn't be registered at the first place, and we won't reach this point.

The webhook is enabled regardless of if single step is on :) because of the labels and `mutate_unlabeled` configs.

adel121 replied:
hmmmm, this is confusing to me 🤔

The wehook name is auto instrumentation, and yet, it is enabled regardless of if single step instrumentation is on? 

If single step instrumentation is off, what would the webhook do?

adel121 replied:
I also see that the webhook is enabled only when `admission_controller.auto_instrumentation.enabled` ([here](https://github.com/DataDog/datadog-agent/blob/4c815329255671890607312a2d864bb8185ff761/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L93))

But, this config is always set to true (it is not set in any place in the helm chart, see [here](https://github.com/search?q=repo%3ADataDog%2Fhelm-charts+admission_controller_auto_instrumentation_enabled&type=code))

This means the webhook is always enabled no matter what? 🤔

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +349,94 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+	}
+
 	// If the pod is "injectable" and annotated with libraries to inject, use those.
 	if w.isPodEligible(pod) {
```
Is there a case where`w.isPodEligible(pod)` can return `false` ? 

I thought we were dropping the mutation request and returning directly before reaching the call to extractLibInfo (see [here](https://github.com/DataDog/datadog-agent/blob/6aad87436465e0ee064c9bd860a6507f54a2680f/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L175-L177))

stanistan replied:
`isPodEligible` is a superset of ssiEnabled. It is really checking to see if we should respect the annotations on the pod.

adel121 replied:
Yes but if `isPodEligible(pod)` returns `false`, we will never reach this part of the code because a similar check was made [here](https://github.com/DataDog/datadog-agent/blob/6aad87436465e0ee064c9bd860a6507f54a2680f/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L175-L177).

In other words, my point is that maybe checking `w.isPodEligible(pod)` here is redundant because we are sure it is true

stanistan replied:
I see! Yeah, the contract of `extractLibInfo` is now a little different, it was doing a second eligibility check even though it didn't need to at all.

adel121 reviewed (commented):
left some comments

stanistan committed:
update hopefully clearer comments

stanistan committed:
clean up extraction logic

stanistan commented:
There's an issue here with a race-condition in using a similar subpath on mounting the volume (I'm hitting in testing)

https://github.com/kubernetes/kubernetes/issues/94198

I can verify that the data is there but it breaks during creation.

jonbodner commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -155,10 +163,13 @@ func initContainerName(lang language) string {
 
+// isEnabledInNamespace checks whether this namespace is opted into or out of
+// single step (auto_instrumentation) outside pod-specific annotations.
 func (w *Webhook) isEnabledInNamespace(namespace string) bool {
```
adding comments is always nice. Thanks!

jonbodner reviewed (approved):
LGTM

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +359,124 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+	// libInfoSourceNone is no source provided.
+	libInfoSourceNone libInfoSource = iota
+	// libInfoSourceLibInjection is when the user sets up manual annotations on their pods.
+	libInfoSourceLibInjection
```
nit:

I am not sure, but I think using `libInfoSourcePodAnnotations` is more conveying here. 

Without the comment above, I think it is not very intuitive to guess that this is used when the user sets up manual annotations on their pods.

stanistan replied:
The reason I want to keep this as `libInjection` is that it's the name of the product (and how this ends up being logged in the environment variables and metrics).

It's not _exactly_ that it's just the annotations, but its the annotations with single step being disabled for the namespace-- updating the comment to that effect.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +359,124 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+	libInfoSourceLibInjection
+	// libInfoSourceSingleStepInstrumentation is when we are using the instrumentation config
+	// to determine which libraries to inject.
+	libInfoSourceSingleStepInstrumentation
+	// libInfoSourceSingleStepLanguageDetection is when we use the language detection
+	// annotation to determine which libs to inject.
+	libInfoSourceSingleStepLangaugeDetection
```
Mentioning just in case, 

Even when using language detection to inject libraries, we still depend on the config of the single step instrumentation. So, there is an overlap between these 2. 

Ideally, when everything is stable, I think language detection should become an implementation detail of SSI, and shouldn't be perceived as a separate source. 

For now, it might be okay to make this distinction.

stanistan replied:
Agree- it’s there right now only to disambiguate some metrics we are producing.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +307,46 @@ func getPinnedLibraries(registry string) []libInfo {
+	injectionEnabled bool
 
-	return nil
+	injected bool
```
I don't see this field populated anywhere.

Maybe you wanted to rename it to `injectionEnabled` but forgot to remove it?

It is also a bit confusing to have both `injectionEnabled` and `injected`, because why would one be true while the other one is not.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +307,46 @@ func getPinnedLibraries(registry string) []libInfo {
+			libs = append(libs, string(lib.lang))
+		}
+
+		_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+			Name:  "DD_INSTRUMENTATION_LANGUAGES_DETECTED",
+			Value: strings.Join(libs, ","),
+		})
+		_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+			Name:  "DD_INSTRUMENTATION_LANGUAGE_DETECTION_INJECTION_ENABLED",
+			Value: strconv.FormatBool(l.injectionEnabled),
+		})
```
Are you sure you don't need to propagate whether language detection was enabled or not?

If we don't propagate this information, then what would happen if the injector receives the following:
```
DD_INSTRUMENTATION_LANGUAGES_DETECTED=
DD_INSTRUMENTATION_LANGUAGE_DETECTION_INJECTION_ENABLED=false
```

This can imply one of the following:
- Language detection is not enabled in the first place
- Language detection is enabled, but for some reason, we detected nothing.

Without the knowledge of whether language detection is enabled or not, we will not be able to decide which of the two implications the correct one is. 

If I could convince you, I'd suggest two ways around this:
- Conditionally inject these 2 env vars only when `language_detection.enabled` and `language_detection.reporting.enabled` are `true`.
- Inject a third env var `DD_INSTRUMENTATION_LANGUAGE_DETECTION_ENABLED`

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +359,124 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+	return e
+}
+
+func (e extractedPodLibInfo) useLanguageDetectionLibs() (extractedPodLibInfo, bool) {
```
can you please use named returns to make it clear what the returned boolean represents 🙏🏻

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -329,35 +359,124 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
-		if len(w.pinnedLibraries) > 0 {
-			return w.pinnedLibraries, false
-		}
+		source = libInfoSourceSingleStepInstrumentation
+		languageDetection = w.getLibrariesLanguageDetection(pod)
```
Few Questions:

- Why is the source set here to `libInfoSourceSingleStepInstrumentation` and not to `libInfoSourceSingleStepLangaugeDetection`
- Why are we only checking for language detection results here, why don't we also check if the user has put annotations on the pod?

stanistan committed:
Updates:

1. Making sure that when we do logging we respect language_detection
   configuration flags and can disambiguate between no reporting and
   reporting empty data (and adding tests for this).

2. Ensuring we respect the container name from the language detection
   annotations and only inject the environment variables there.

3. Some refactoring of the container mutators to make this simpler to
   work with.

4. Additions of `containerMutators`  and `libRequirementOptions` to
   capture the additional mutations we're doing.

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -308,15 +307,59 @@ func getPinnedLibraries(registry string) []libInfo {
+	injectionEnabled bool
+}
+
+func (l *libInfoLanguageDetection) containerMutator(v version) containerMutator {
```
So if I understand correctly:

- if language detection is not enabled, we inject nothing
- if language detection is enabled we inject what is detected (even if nothing was detecteD) + a flag indicating if injection is enabled or not.

Correct?

stanistan replied:
Yep!

adel121 commented on pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:
```diff
@@ -185,8 +196,8 @@ func (w *Webhook) inject(pod *corev1.Pod, ns string, _ dynamic.Interface) (bool,
-	if len(libsToInject) == 0 {
+	extractedLibInfo := w.extractLibInfo(pod)
+	if len(extractedLibInfo.libs) == 0 {
 		return false, nil
```
If this line is hit, the following 2 injections (security client & profiling client) will be ignored. 

Please make sure if this is the intended behaviour (i.e. security client config & profiling client config should not be injected if APM injected no tracing libraries)

stanistan replied:
Will check-- this PR isn't changing this behavior though-- prior behavior would _also_ have aborted injection in this case.

If it's not intended, we can fix in a follow up.

adel121 reviewed (approved):
LGTM for container-platform files

stanistan commented:
/merge
